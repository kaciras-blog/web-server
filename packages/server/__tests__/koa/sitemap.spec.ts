import { expect, it, vi } from "vitest";
import Koa from "koa";
import supertest from "supertest";
import sitemapMiddleware from "../../lib/koa/sitemap";

const mockFetch = vi.fn();
global.fetch = mockFetch;

const DATA = {
	items: [
		{ id: 0, update: 1572446089218, urlTitle: "test-0" },
		{ id: 123, update: 1572446089218, urlTitle: "test-1" },
	],
};

const app = new Koa();
app.use(sitemapMiddleware("123.45.67.89"));
const callback = app.callback();

mockFetch.mockResolvedValue({ status: 200, json: () => DATA });

it("should response sitemap xml", async () => {
	await supertest(callback)
		.get("/sitemap.xml")
		.expect(200)
		.expect((res) => expect(res.text).toMatchSnapshot());
});

it("should output short date for Baidu", async () => {
	await supertest(callback)
		.get("/sitemap.xml")
		.query({ type: "baidu" })
		.expect(200)
		.expect((res) => expect(res.text).toMatchSnapshot());
});

it("should recognize Baidu spider by User-Agent", async () => {
	function _test(userAgent: string) {
		return supertest(callback)
			.get("/sitemap.xml")
			.set("User-Agent", userAgent)
			.expect(200)
			.expect((res) => expect(res.text).toMatchSnapshot());
	}

	await _test("Mozilla/5.0 (compatible; Baiduspider/2.0; +http://www.baidu.com/search/spider.html)");

	await _test("Mozilla/5.0 (Linux;u;Android 4.2.2;zh-cn;) AppleWebKit/534.46 (KHTML,like Gecko) " +
		"Version/5.1 Mobile Safari/10600.6.3 (compatible; Baiduspider/2.0; +http://www.baidu.com/search/spider.html)");
});

it("should return 503 on error", async () => {
	mockFetch.mockRejectedValue(new Error());
	await supertest(callback).get("/sitemap.xml").expect(503);
});
