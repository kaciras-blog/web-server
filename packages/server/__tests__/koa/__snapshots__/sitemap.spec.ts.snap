// Vitest Snapshot v1

exports[`should output short date for Baidu 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://blog.kaciras.com/article/0/test-0</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url><url><loc>https://blog.kaciras.com/article/123/test-1</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url></urlset>"`;

exports[`should recognize Baidu spider by User-Agent 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://blog.kaciras.com/article/0/test-0</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url><url><loc>https://blog.kaciras.com/article/123/test-1</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url></urlset>"`;

exports[`should recognize Baidu spider by User-Agent 2`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://blog.kaciras.com/article/0/test-0</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url><url><loc>https://blog.kaciras.com/article/123/test-1</loc><lastmod>2019-10-30</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url></urlset>"`;

exports[`should response sitemap xml 1`] = `"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?><urlset xmlns=\\"http://www.sitemaps.org/schemas/sitemap/0.9\\" xmlns:news=\\"http://www.google.com/schemas/sitemap-news/0.9\\" xmlns:xhtml=\\"http://www.w3.org/1999/xhtml\\" xmlns:image=\\"http://www.google.com/schemas/sitemap-image/1.1\\" xmlns:video=\\"http://www.google.com/schemas/sitemap-video/1.1\\"><url><loc>https://blog.kaciras.com/article/0/test-0</loc><lastmod>2019-10-30T14:34:49.218Z</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url><url><loc>https://blog.kaciras.com/article/123/test-1</loc><lastmod>2019-10-30T14:34:49.218Z</lastmod><changefreq>monthly</changefreq><priority>1.0</priority></url></urlset>"`;
